test/lib/sdkdriver/test/unit/spec/transport/websocket/client.ts

```diff
@@ -0,0 +1,157 @@
+import WSClientTransport from '../../../../../src/transport/websocket/client';
+import { WebSocket } from '../../../mocks';
+import * as assert from 'assert';
+import * as sinon from 'sinon';
+
+describe('WSClientTransport', () => {
+  describe('#close', () => {
+    let closeEventPromise: Promise<void>;
+    let transport: WSClientTransport;
+    let wsClient: WebSocket;
+
+    before(async () => {
+      transport = new WSClientTransport('foo', { WebSocket });
+      const openPromise: Promise<void> = transport.open();
+      wsClient = transport._wsClient;
+      wsClient.readyState = WebSocket.OPEN;
+      wsClient.dispatchEvent({ type: 'open' });
+      await openPromise;
+      closeEventPromise = new Promise(resolve => transport.on('close', resolve));
+      transport.close();
+    });
+
+    it('should call .close on the underlying WebSocket', async () => {
+      sinon.assert.calledOnce(wsClient.close);
+    });
+
+    it('should emit "close" on the WSClientTransport', () => {
+      return closeEventPromise;
+    });
+  });
+
+  describe('#open', () => {
+    let transport: WSClientTransport;
+
+    beforeEach(() => {
+      transport = new WSClientTransport('foo', { WebSocket });
+    });
+
+    it('should return a Promise', () => {
+      assert(transport.open() instanceof Promise);
+    });
+
+    it('should create a new WebSocket with the url passed to the constructor', () => {
+      transport.open();
+      sinon.assert.calledWith(WebSocket, 'foo');
+    });
+
+    [
+      {event: 'open', readyState: 'OPEN'},
+      {event: 'close', readyState: 'CLOSED'}
+    ].forEach(({ event, readyState }) => {
+      context(`when the WebSocket ${event === 'open' ? 'opens' : 'fails to open'}`, () => {
+        let promise: Promise<void>;
+        let wsClient: any;
+
+        beforeEach(() => {
+          transport._sendBuffer.push('foo');
+          transport._sendBuffer.push('bar');
```

Comment:
I'm surprised TypeScript allows this. `_sendBuffer` is private. Can we set this test up in a way where we rely only on public API?